tidy3d/plugins/waveguide/rectangular_dielectric.py

```diff
@@ -0,0 +1,560 @@
+        units=MICROMETER,
+    )
+
+    sidewall_medium: MediumType = pydantic.Field(
+        None,
+        description="Medium associated with the sidewall layer to model sidewall losses.",
+    )
+
+    surface_thickness: float = pydantic.Field(
+        0.0,
+        description="Thickness of the surface layers defined on the top of the waveguide and  "
+        "slab regions (if any).",
+        units=MICROMETER,
+    )
+
+    surface_medium: MediumType = pydantic.Field(
+        None,
+        description="Medium associated with the surface layer to model surface losses.",
+    )
+
+    origin: Coordinate = pydantic.Field(
+        (0, 0, 0),
+        description="Center of the waveguide geometry.  This coordinate represents the base "
+        "of the waveguides (substrate surface) in the normal axis, and center of the geometry "
+        "in the remaining axes.",
+        units=MICROMETER,
+    )
+
+    length: Size1D = pydantic.Field(
+        1e30,
+        description="Length of the waveguides in the propagation direction",
+        units=MICROMETER,
+    )
+
+    propagation_axis: Axis = pydantic.Field(
+        0,
+        description="Axis of propagation of the waveguide",
+    )
+
+    normal_axis: Axis = pydantic.Field(
+        2,
+        description="Axis normal to the substrate surface",
+    )
+
+    mode_spec: ModeSpec = pydantic.Field(
+        ModeSpec(),
+        description=":class:`ModeSpec` defining waveguide mode properties.",
+    )
+
+    grid_resolution: int = pydantic.Field(
+        15,
+        description="Solver grid resolution per wavelength.",
+    )
+
+    max_grid_scaling: float = pydantic.Field(
+        1.2,
+        description="Maximal size increase between adjacent grid boundaries.",
+    )
+
+    @pydantic.validator("wavelength", "core_width", "gap", always=True)
+    def _set_array(cls, val):
+        if isinstance(val, float):
+            return numpy.array((val,))
+        return numpy.array(val)
+
+    @pydantic.validator("box_medium", always=True)
+    def _set_box_medium(cls, val, values):
+        return values["clad_medium"] if val is None else val
+
+    @pydantic.validator("clad_thickness", always=True)
+    def _set_clad_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["clad_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("box_thickness", always=True)
+    def _set_box_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["box_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
```

Comment:
might be simpler and clearer to let `side_margin` (and other similar things) be `@cached_property`s

Reply:
Is this what you mean?
```python
    @cached_property
    def clad_thickness(self):
        if "clad_thickness" in self.__fields_set__:
            return self.__fields_set__["clad_thickness"]
        wavelength = self.wavelength
        medium = self.clad_medium
        n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
        lda = wavelength / n
        return 1.5 * lda.max()
```

Reply:
do we intend to set clad_thickness directly in the plugin? if so ignore what I said about property. What I was getting at is that if the clad thickness is just a derived property, we should do it as a property instead of as a validator